feature/timezone

JStaegerino opened the pull request:
The Home Assistant time zone is now included in the calculation. This means that the script can be used anywhere. To achieve this, it was necessary to switch from Python_Script to pyscript, as it was not possible to import the time zone in Python_script.

JStaegerino committed:
feat: add rotation of weekdays

Adds a configuration option to start the calendar week on either Sunday or Monday.

Improves flexibility by allowing users to choose their preferred day to begin the week. Refactors weekday calculation and display logic to accommodate both Sunday and Monday as the first day.

JStaegerino committed:
feat: add timezone from homeassistant

Migrates calendar data conversion from python_scripts to pyscript for better integration.

This change provides a service to convert Home Assistant calendar data into a format suitable for display on ESPHome devices. It normalizes event data, handles all-day events correctly, and sorts events by date and time. It also extracts relevant information like location names and ensures proper encoding of event summaries. Additionally, it sets sensor attributes for day name and month for localization purposes. The script uses the time zone from Home Assistant and calculates the times of the appointments accordingly.

JStaegerino committed:
Migrates calendar data conversion to pyscript

Changes the calendar data conversion from a `python_script` to a `pyscript` for better integration and maintainability.

Updates the documentation to reflect the change and clarify the `pyscript` integration requirement.

JStaegerino committed:
feat: fix some comments

sannong commented on HomeAssistant_sensor/HA_epaper.yaml:
```diff
@@ -25,49 +25,32 @@ template:
   # Calendar        
   - trigger:
       - platform: time_pattern
```
It looks like the newest HA version wants this to be "trigger" instead of "platform" to conform with the latest syntax. Otherwise it gives "String does not match the pattern of "LEGACY_SYNTAX^"". This applies to a couple other places as indicated as well.

sannong commented on HomeAssistant_sensor/HA_epaper.yaml:
```diff
@@ -25,49 +25,32 @@ template:
-          {% set delta = ((event - now().timestamp()) / 86400) | round (1) %}
-          {{ -delta }}
-        state_class: measurement
+      - service: calendar.get_events
```
Latest HA wants "action" instead of "service"

sannong commented on HomeAssistant_sensor/HA_epaper.yaml:
```diff
@@ -25,49 +25,32 @@ template:
+          duration:
+            days: 7
+        response_variable: calendar_response
+      - service: pyscript.esp_calendar_convert
```
"action" instead of "service"

sannong commented on HomeAssistant_sensor/HA_epaper.yaml:
```diff
@@ -25,49 +25,32 @@ template:
-        state_class: measurement
+      - service: calendar.get_events
+        target:
+          entity_id:
```
Is there a reason you removed the comments from the original code? I kept them in just to help people see where to add/change things and what steps were being taken. This applies to a few items.

sannong commented on HomeAssistant_sensor/HA_epaper.yaml:
```diff
@@ -25,49 +25,32 @@ template:
   - trigger:
       - platform: time_pattern
-        hours: /1
-      - trigger: homeassistant
```
Can you leave in the homeassistant start trigger as well so the calendar data is fetched when homeassistant restarts? Otherwise I think it has to wait for the hour trigger to catch and may have stale/unpopulated data?

sannong commented on README.md:
```diff
@@ -62,9 +62,10 @@ This project uses the following integrations:
+### Pyscript
 
-A python script is used to format the calendar event data. Copy the python_scripts folder and contents provided to your Home Assistant root directory (creating the python_scripts folder at the root level if needed).
+A python script is used to format the calendar event data. Copy the pyscripts folder and contents provided to your Home Assistant root directory (creating the pyscripts folder at the root level if needed).
```
Should be "pyscript" folder not "pyscripts" folder since that is the actual name.

sannong commented on HomeAssistant_sensor/HA_epaper.yaml:
```diff
@@ -25,49 +25,32 @@ template:
   # Calendar        
```
I'm having trouble getting this to work. I had to uncomment the "python_script" line above this or else HACKs threw an error and wouldn't load. When I do that it loads but I don't get any output from the python script as far as I can tell and I get this error in the HASS log, From what I can tell the python script runs (I added a log message to it's main routine and do see that message come out), but it doesn't seem to return data?

2025-09-20 23:30:36.790 WARNING (MainThread) [py.warnings] /config/custom_components/pyscript/eval.py:1983: RuntimeWarning: coroutine 'EvalFuncVar.__call__' was never awaited
  return func(*args, **kwargs)

2025-09-20 23:30:36.790 ERROR (MainThread) [custom_components.pyscript.file.esp_calendar_data_conversion.esp_calendar_convert] esp_calendar_convert error: cannot unpack non-iterable object

Does this work for you as submitted? I copied the python script exactly as in this PR.

sannong commented on HomeAssistant_sensor/HA_epaper.yaml:
```diff
@@ -25,49 +25,32 @@ template:
+          calendar: "{{ calendar_response }}"
+          now: "{{ now().date() }}"
+          entity_id: sensor.esp_calendar_data
+    - sensor:
```
Indentation of this section appears to be incorrect when I paste, needs to be left indented one more place

JStaegerino commented:
Hey @JStaegerino, while reviewing your PR, I'd suggest the following code changes:

👉 [**Code Suggestion for #5**](https://gitkraken.dev/drafts/fc5681d0-17bb-4e3f-bd0e-1a9546768d81?source=PRComment)
>https://github.com/sannong/eink_weather/pull/5

You can also review and apply these suggestions [locally on your machine](https://gitkraken.dev/link/drafts/fc5681d0-17bb-4e3f-bd0e-1a9546768d81?type=suggested_pr_change&prEntityId=WyJnaXRodWIiLCJwciIsIjEiLCIiLCJQUl9rd0RPT2pqZUZzNnBxUEk2Il0=&source=PRComment).

<details><summary>Learn more about GitKraken Code Suggest</summary>

>[Code Suggest](https://gitkraken.dev/redirect/aHR0cHM6Ly93d3cuZ2l0a3Jha2VuLmNvbS9zb2x1dGlvbnMvY29kZS1zdWdnZXN0P3NvdXJjZT1naXRodWIuY29tJm1lZGl1bT1wcl9jb21tZW50?source=PRSuggest&event=redirectToCodeSuggestPage&provider=github) liberates your code reviews from GitHub's restrictive, comment-only feedback style. As simple as suggesting changes in a Google-doc, provide real code suggestions from where you code, e.g. your IDE, and on anything in your project — not just on the lines of code changed in the PR.

_Join your team on [GitKraken](https://gitkraken.dev/redirect/aHR0cHM6Ly93d3cuZ2l0a3Jha2VuLmNvbT9zb3VyY2U9Z2l0aHViLmNvbSZtZWRpdW09cHJfY29tbWVudA==?source=PRSuggest&event=redirectToIndexPage&provider=github) to speed up PR review._
</details>

JStaegerino committed:
feat: corrected a few blocks

JStaegerino commented:
You couldn't make this up... I have now reprogrammed the ESP (deep sleep, as I am working with a battery). Now I have the same errors in the pyscript...
I am in the process of looking for a solution.

JStaegerino committed:
feat: calendar data conversion script

This commit introduces a new script for converting calendar data into a format suitable for use with ESP devices.

The script includes functions for:
- Parsing and localizing datetime strings
- Normalizing event shapes
- Sorting and aggregating events by date
- Providing service to convert calendar data

The script also handles all-day events and provides the closest end time for events happening today.

sannong commented:
Thanks for the update! I wonder why you weren't seeing an error before like I was. Let me know if you need me to try anything.
